### PR TITLE
Fix #3919: Marquee for toolbar titles make  RTL compatible.

### LINKED ISSUE #3919: The marquee for toolbar titles has to be RTL compatible.
The marquee for toolbar titles goes left-to-right in RTL device. The actual behavior has to be from right-to-left for RTL device.

### PR DESCRIPTION
**Pull request by veena14cs**
## Explanation

This PR fixes #3919. This PR fixes Marquee direction for toolbar titles to from right-to-left in RTL devices.

https://user-images.githubusercontent.com/16301028/137023878-e0b8fda5-ed80-4182-b268-131608d74d68.mp4

https://user-images.githubusercontent.com/16301028/137025589-b0b9fae5-ed6b-4105-ad61-fbcf453176f3.mp4

**Device**
Nexus 5 Kitkat

https://user-images.githubusercontent.com/16301028/137274925-00458346-8e80-4fb7-b719-7d14cd09ddf0.mp4

https://user-images.githubusercontent.com/16301028/137276845-98120610-d07b-476f-82ae-6cf42dddde16.mp4

## Screenshots of Testcases after RTL 

Screenshot (15)

Screenshot (16)

Screenshot (17)

Screenshot (22)

Screenshot (27)

##  Screenshots of Testcases before RTL 

Screenshot (18)

Screenshot (19)

Screenshot (21)

Screenshot (23)

Screenshot (26)

## Essential Checklist

- [ ] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only

If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-(A11y)-Guide))
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

### CODE REVIEW

File: app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt]: Mis-reformatting?

File: app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt]: This seems really error-prone (it's super easy to forget this when creating or changing activities). Would it be reasonable to create our own toolbar subclass that sets up its text direction automatically?

I'm also a bit confused. Why doesn't this just automatically work in the system?

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @veena14cs. If this doesn't work in an emulator, how have you verified that it's correct? It was my understanding that the built-in marquee in AndroidX's toolbar doesn't properly support RTL. Is that not the case?

**Comment by veena14cs** [comment]
> Thanks @veena14cs. If this doesn't work in an emulator, how have you verified that it's correct? It was my understanding that the built-in marquee in AndroidX's toolbar doesn't properly support RTL. Is that not the case?

By default it is not supporting. If the textview direction is set to RTL then the marquee direction goes right-to-left.

File: app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt
Comment (veena14cs) [reply app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt]: Done.

File: app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt]: Created custom marquee toolbar textview for setting textdirection. It is not setting the marquee direction by default when device is in RTL mode. We need to set textdirection to get correct Marquee effect.

**Comment by veena14cs** [comment]
> Thanks @veena14cs. If this doesn't work in an emulator, how have you verified that it's correct? It was my understanding that the built-in marquee in AndroidX's toolbar doesn't properly support RTL. Is that not the case?

It is working on emulator now I will be adding up the screen recording. Initially I think I added it in xml or I mistook it that I couldn't observe the correct behavior on the emulator when I was toggling between the Languages and playing around the same screen without reloading it.

File: app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt]: Why is this needed?

File: app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt]: Just to clarify, do we not initially play the marquee since this only changes direction upon clicking? Also, does this result in any stuttering when the user taps on the text view?

File: app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt]: Why not use 'this' instead of passing in a TextView?

File: app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt]: Is this something that we can do in onAttachedToWindow instead of onClick?

File: app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt]: I'm wondering if this is actually correct. I'm assuming we actually want to set the text direction using https://github.com/oppia/oppia-android/blob/f266da8dc41c7313b1e220f5eed38d05461cc078/utility/src/main/java/org/oppia/android/util/locale/OppiaLocale.kt#L241. I think we inject this in other views, so see usages of that method to see how to get access to a DisplayLocale.

Without using DisplayLocale, this might get set based on the system locale rather than the app-specified one.

File: app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt]: Why is this needed? Can we just set the direction to the one we know it needs to be per my comment below?

File: app/src/sharedTest/java/org/oppia/android/app/customview/MarqueeToolbarTextViewTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/android/app/customview/MarqueeToolbarTextViewTest.kt:159]: Prefer directly importing withId()/click().

File: app/src/main/res/layout/exploration_activity.xml
Comment (BenHenning) [inline app/src/main/res/layout/exploration_activity.xml:41]: Should we be removing the marquee-specific properties from these uses & instead specify them directly in the custom view? Otherwise it's sort of not a marquee text view if we need to set up the marquee properties each place that it needs to be used.

File: app/BUILD.bazel
Comment (BenHenning) [inline app/BUILD.bazel]: It doesn't import org.oppia.android.R, so it should go in the other VIEW list.

File: app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt]: Should be on one line. Ditto for all other similar lines elsewhere in the PR.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @veena14cs. Per the videos, I think this looks quite god. Just had some comments about the latest approach, but this seems like a good approach.

Can you please check this on a KitKat emulator to make sure that it's working correctly for KitKat, too? To test, you should use the Bazel build (which you can download from CI: https://github.com/oppia/oppia-android/pull/3920/checks?check_run_id=3875395434 see artifacts)--I'm not sure whether the Gradle build works on KitKat.

The Bazel failure in CI seems to be an infra flake. I restarted it.

File: app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt]: Removed this.

File: app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt]: Removed this method from onclick and moved to onAttachWindow().

File: app/src/sharedTest/java/org/oppia/android/app/customview/MarqueeToolbarTextViewTest.kt
Comment (veena14cs) [reply app/src/sharedTest/java/org/oppia/android/app/customview/MarqueeToolbarTextViewTest.kt:159]: Done.

File: app/src/main/res/layout/exploration_activity.xml
Comment (veena14cs) [reply app/src/main/res/layout/exploration_activity.xml:41]: Done.

File: app/BUILD.bazel
Comment (veena14cs) [reply app/BUILD.bazel]: Done.

File: app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt
Comment (veena14cs) [reply app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt]: Done.

File: app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt]: Done.

File: app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt]: Done.

File: app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt]: Done.

File: app/src/sharedTest/java/org/oppia/android/app/customview/MarqueeToolbarTextViewTest.kt
Comment (veena14cs) [reply app/src/sharedTest/java/org/oppia/android/app/customview/MarqueeToolbarTextViewTest.kt:159]: Done.

File: app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt]: Removed this function.

File: app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt:28]: Can we keep the previous constructor declaration (as we do for other custom views)?

File: app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt:52]: Suggest just inlining this isRtl value to avoid potentially accessing it before it's actually initialized (this is a bit clearer when trying to access injectable fields, but the by lazy is once removed & a bit harder to tell when it can be called).

File: app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt]: Could we maybe just do something like ``textDirection = resourceHandler.getLayoutDirection()`` or do the constants differ?

File: app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt:40]: Is the '20' here in dp or pixels? We might need to convert for it to be correct.

File: app/src/main/res/layout/marquee_toolbar_text_view_test_fragment.xml
Comment (BenHenning) [inline app/src/main/res/layout/marquee_toolbar_text_view_test_fragment.xml:2]: We aren't using databinding to inflate this view--should this top-level declaration instead be the ConstraintLayout declared below?

File: app/src/main/res/layout/story_fragment.xml
Comment (BenHenning) [inline app/src/main/res/layout/story_fragment.xml:46]: Is this one defined programmatically in the custom view? It seems missing. Is it default true for clickable text views?

File: app/src/sharedTest/java/org/oppia/android/app/customview/MarqueeToolbarTextViewTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/android/app/customview/MarqueeToolbarTextViewTest.kt:136]: Is this correct? It should pass on Robolectric, but might not on Gradle. See that other tests are restricted to running on Bazel--can you see if this passes on Bazel before disabling this for Robolectric?

File: app/src/sharedTest/java/org/oppia/android/app/customview/MarqueeToolbarTextViewTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/android/app/customview/MarqueeToolbarTextViewTest.kt]: Please also import withId directly.

And, this can fit on one line (wrapping isn't needed).

File: app/src/sharedTest/java/org/oppia/android/app/topic/TopicFragmentTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/android/app/topic/TopicFragmentTest.kt:154]: This shouldn't be removed--it stops the test from running.

File: app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt]: Why remove these tests? Ditto for ExplorationActivityTest and RevisionCardActivityTest.

File: scripts/assets/kdoc_validity_exemptions.textproto
Comment (BenHenning) [inline scripts/assets/kdoc_validity_exemptions.textproto:31]: Don't exempt this--prefer to add the missing documentation, instead.

File: app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt:16]: ``Custom text view that should be configured as the child of a toolbar to display a toolbar's title with an RTL-capable marquee effect.``

Bit more specific on why & how to use this.

File: app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt
Comment (BenHenning) [reply app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt]: Thanks, but I don't think that actually answers my questions about juttering/initial behavior.

I think I have the answer to the first one (that tapping is necessary), but can you clarify on the second point: is there any juttering when in RTL mode? It doesn't seem like it from the video, but the video might not show jutters.

File: app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt
Comment (BenHenning) [reply app/src/main/java/org/oppia/android/app/customview/MarqueeToolbarTextView.kt]: Replaced with newer comment (I don't think this was fully resolved).

File: app/src/sharedTest/java/org/oppia/android/app/customview/MarqueeToolbarTextViewTest.kt
Comment (BenHenning) [reply app/src/sharedTest/java/org/oppia/android/app/customview/MarqueeToolbarTextViewTest.kt:159]: One was done, but not the other. Replaced with a newer comment.

File: app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt
Comment (BenHenning) [reply app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt]: Not fully resolved, but replaced with a newer comment.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @veena14cs. Took another full pass--PTAL.

File: app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt
Comment (veena14cs) [reply app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt]: I removed it because I need to create Fragment test for these activity test as its not present in current code, so I just removed it from here.

**Comment by rt4914** [comment]
For this particular PR I will have to defer to @BenHenning . Once he approves I will take a look at my code-owner files just to be sure.

**Comment by veena14cs** [comment]
Closing this PR as its getting fixed in #3935

**Pull request closed without merging**
